### ISSUE TITLE
Text positioning api

### ISSUE THREAD

**Issue by pvdz** [uxebu/bonsai#84]
``` js
new Text('foo').attr({x,y,width:500,height:300,valign:'top',halign:'middle'})
```

There would be two new attributes: `halign` and `valign` which control horizontal and vertical alignment respectfully. Valid values for horizontal would be `'left','middle','right'`, and vertical `'top','middle','bottom'` (`middle` comes from css, i think it'd be nice if `center` would also map to `middle` here).

I know text has way more vertical lines to align to, but these three are the most important bits.

The alignment should (probably?) be relative to the bounding box of the `Text` element. So if you specify a `Text` with `width=500` and `halign='right'` then the _right-hand side_ of the text should be at 500px from the left-hand-side of the text element. Might be obvious but just pointing it out anyways :)

**Comment by basecode** [comment]
Good point. Related: https://github.com/uxebu/bonsai/tree/text-align

**Comment by there4** [comment]
+1

I've also noticed that the text api also seems to always return zeroes for the getBoundingBox. It appears that this is not implemented, and is just returning the parent display object method.

**Comment by basecode** [comment]
@there4, you're right. The reason for that is that it isn't trivial. I would suggest to let the author (manually) define a TextBox (x, y, width, height) instead.

It's not trivial because:
- There is no API (e.g. DOM) available in the runner context (where all bonsai movies are executed) that could do the job for us
- We don't have the numbers. A glyph's outline depends on the inner-bbox, hinting, kerning, ligatures… We don't have access to these numbers in case of system fonts.
- Even if we would have access to the font's source-file, it would probably take some time to get the numbers correct.

Feel free contributing!

**Comment by iammerrick** [comment]
I'm sorry the documentation for this piece is a bit sparse I am struggling to figure out how to simply accomplish this... :-)
![Screen Shot 2013-04-22 at 11 22 01 AM](https://f.cloud.github.com/assets/109635/410545/898d5f60-ab71-11e2-82f6-468b3f80c900.png)

**Comment by basecode** [comment]
> I'm sorry the documentation for this piece is a bit sparse I am struggling to figure out how to simply accomplish this... :-)

It's sparse because textAlign is not supported so far. I created a pull-request, see #208.

**Comment by washu** [comment]
could you not after the svg render draw the text/textspan, grab the bounding box to return the width,height back to the runner to atleast populate the getBoundingBox call?

**Comment by basecode** [comment]
> could you not after the svg render draw the text/textspan, grab the bounding box to return the width,height back to the runner to atleast populate the getBoundingBox call?

Yes, we could do that. 3 reasons why we didn't:
- Expensive: As you already mentioned requesting `getBoundingBox` would imply sending a message to the renderer and back to the runner. In cases where a user does this multiple times within a frame (also per frame) it's getting expensive. Also think of a runner that is executed on a server instead of a worker or iFrame.
- API: We would need to make `getBoundingBox` async.

```
text.getBoundingBox(function(computedBoundingBoxAfterSomeRoundtripsBetweenRunnerAndRenderer) {
 …
});
```
- Stage: We would be forced to to the "expensive" round-trip even when the text wasn't even added to the stage. We try to avoid sending messages to the renderer for DisplayObjects that aren't supposed to be rendered.

**Comment by washu** [comment]
well that would be more of the extreme side, i was thinking to only send an update out from the runner after the svg is rendered, and send an update back out. so the calling Text bject would be updated only after it was rendered. not an async call for every time getBoundingBox was called. that should only return the last known bounding box render state. for height and width.

**Comment by basecode** [comment]
Yes, caching makes sense. But the API call remains unbound to rendering and is also still async. I'm totally ok if you would offer a pull-request that proves me wrong of course. :)

**Comment by basecode** [comment]
There has been an interesting [discussion on irc](http://echelog.com/logs/browse/bonsaijs/1370556000) regarding Text's bounding box and the current state of text wrapping in SVG and CSS.
